amethyst_tiles/src/morton/mod.rs

```diff
@@ -0,0 +1,231 @@
+            len: x * y,
+        }
+    }
+
+    #[inline]
+    fn encode(&self, x: u32, y: u32, z: u32) -> Option<u32> {
+        use bitintr::Pdep;
+
+        #[cfg(debug_assertions)]
+        {
+            let check = u32::max_value() / 3;
+            if x > check || y > check || z > check {
+                return None;
+            }
+        }
+
+        let morton = (x.pdep(0x5555_5555) | y.pdep(0xAAAA_AAAA)) + (z * self.len);
+
+        Some(morton)
+    }
+    #[inline]
+    fn decode(&self, mut morton: u32) -> Option<(u32, u32, u32)> {
+        use bitintr::Pext;
+
+        let z = morton / self.len;
+        morton -= z * self.len;
+
+        Some((morton.pext(0x5555_5555), morton.pext(0xAAAA_AAAA), z))
+    }
+}
+
+#[inline]
+#[cfg(not(target_feature = "bmi2"))]
+pub fn encode(x: u32, y: u32, z: u32) -> u32 {
+    morton_encode_lut(x, y, z)
+}
+
+#[inline]
+#[cfg(not(target_feature = "bmi2"))]
+pub fn decode(morton: u32) -> (u32, u32, u32) {
+    morton_decode_lut(morton)
+}
+
+#[inline]
+pub fn min(morton1: u32, morton2: u32) -> u32 {
+    let d1 = decode(morton1);
+    let d2 = decode(morton2);
+
+    if d1.0 > d2.0 || d1.1 > d2.1 || d1.2 > d2.2 {
+        morton2
+    } else {
+        morton1
+    }
+}
+
+#[inline]
+pub fn max(morton1: u32, morton2: u32) -> u32 {
+    let d1 = decode(morton1);
+    let d2 = decode(morton2);
+
+    if d1.0 < d2.0 || d1.1 < d2.1 || d1.2 < d2.2 {
+        morton2
+    } else {
+        morton1
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use more_asserts::*;
+    use rayon::prelude::*;
+
+    #[test]
+    fn morton_intr_decode_encode_match() {
+        let test_side: u32 = 128; // 12-bit?
+
+        (0..test_side).into_par_iter().for_each(|x| {
+            for y in 0..test_side {
+                for z in 0..test_side {
+                    let morton = morton_encode_intr_3d(x, y, z);
+                    let decode = morton_decode_intr_3d(morton);
+                    assert_eq!((x, y, z), decode);
+                }
+            }
+        });
+    }
+
+    #[test]
+    fn morton_intr_match_lut() {
+        let test_side: u32 = 128;
+        (0..test_side).into_par_iter().for_each(|x| {
+            for y in 0..test_side {
+                for z in 0..test_side {
+                    let morton_lut = morton_encode_lut(x, y, z);
+                    let morton_intr = morton_encode_intr_3d(x, y, z);
+                    assert_eq!(morton_lut, morton_intr);
+                }
+            }
+        });
+    }
+
+    #[test]
+    fn morton_within_array() {
+        let test_side: u32 = 128; // 12-bit?
+        let max: u32 = test_side * test_side * test_side;
+
+        (0..test_side).into_par_iter().for_each(|x| {
+            for y in 0..test_side {
+                for z in 0..test_side {
+                    let morton = morton_encode_lut(x, y, z);
+                    assert_lt!(morton, max);
+                }
+            }
+        });
+    }
```

Comment:
This file could definitely use some tests for `min`, `max`, and probably `cmp` if it's going to be added.

Reply:
Tests added